[wallet] Clarify wallet initialization / destruction interface

jnewbery opened the pull request:
Apologies for the mostly code move only PR. This is a pre-req for both #10740 and #10762 

All wallet component initialization/destruction functions are now in their own `wallet/init.cpp` translation unit and are no longer static functions on the CWallet class. The bitcoin_server also no longer has any knowledge that there are multiple wallets in vpwallet.

There should be no changes in behavior from this PR.

promag commented on src/init.cpp:
```diff
@@ -187,11 +187,6 @@ void Shutdown()
     StopREST();
     StopRPC();
     StopHTTPServer();
-#ifdef ENABLE_WALLET
```
Nothing replacing this code?

jnewbery replied:
Correct - I don't believe the early wallet flush is necessary (although I may be wrong). See PR description:

> The most important one is that Flush() no longer gets called twice on shutdown. I think this is fine, and all the tests pass, but this probably requires some careful consideration from people who are more familiar with the wallet code.

promag commented on src/init.cpp:
```diff
@@ -230,9 +225,7 @@ void Shutdown()
         pblocktree = NULL;
     }
 #ifdef ENABLE_WALLET
```
What about moving these `#ifdef ENABLE_WALLET` to `walletinit.cpp`?

jnewbery replied:
Because `walletinit.cpp` is in libbitcoin_wallet, so is not linked if `ENABLE_WALLET` isn't defined. #10762 will remove the libbitcoin_server -> libbitcoin_wallet dependencies entirely from init.cpp.

practicalswift commented:
Rebase needed :-)

jnewbery commented:
This isn't going in until after 0.15. which includes some wallet changes. I'll rebase after 0.15 has been cut.

practicalswift commented:
@jnewbery Got it! Thanks for the clarification.

jnewbery commented:
This overlaps significantly with #10976, which is easier to review since it's mostly move-only.

I'll rebase this on #10976 once that gets merged.

jnewbery committed:
[wallet] Add FlushWallets() function to wallet/init.cpp

jnewbery committed:
[wallet] Add CloseWallets() function to wallet/init.cpp

jnewbery committed:
[wallet] Add StopWallets() function to wallet/init.cpp

jnewbery committed:
[wallet] Add RegisterWalletRPC() function to wallet/init.cpp

jnewbery commented:
Rebased on @ryanofsky's #10976 and reduced scope. This PR now only moves the remaining wallet startup/shutdown functions to `wallet/init.cpp` in preparation for #10740 (dynamic loading/unloading of wallets) and #10762 (remove wallet dependencies from init.cpp).

There should be no change in behavior from this PR.

ryanofsky reviewed (commented):
utACK 3f425ee4263f66ab039c089d189b6cad0abbc23c. Looks like a simple cleanup with no change in behavior. FWIW I have similar functions in #10973, but am using {start, stop, shutdown} instead of {WalletCompleteStartup, FlushWallets, ShutdownWallets}. Links: [declarations](https://github.com/ryanofsky/bitcoin/blob/pr/wipc-sep/src/ipc/interfaces.h), [definitions](https://github.com/ryanofsky/bitcoin/blob/pr/wipc-sep/src/ipc/local/bitcoin-walletd.cpp)

promag commented on src/wallet/init.h:
```diff
@@ -6,14 +6,22 @@
 #ifndef BITCOIN_WALLET_INIT_H
 #define BITCOIN_WALLET_INIT_H
 
+#include "wallet/rpcwallet.h"
```
Remove?

promag commented on src/wallet/init.cpp:
```diff
@@ -245,3 +253,28 @@ bool InitLoadWallet()
 }
+
+void WalletCompleteStartup(CScheduler& scheduler) {
+    for (CWalletRef pwallet : vpwallets) {
```
Offtopic, we could ditch `CWalletRef`.

promag commented on src/init.cpp:
```diff
@@ -265,10 +265,7 @@ void Shutdown()
-        delete pwallet;
-    }
-    vpwallets.clear();
+    DetachWallets();
```
👍 `FreeWallets`.

promag commented on src/wallet/init.h:
```diff
@@ -22,4 +30,15 @@ bool WalletVerify();
+void ShutdownWallets();
+
+//! Detach all wallets
+void DetachWallets();
```
Newline after.

promag commented on src/wallet/init.h:
```diff
@@ -22,4 +30,15 @@ bool WalletVerify();
+//! Shutdown all wallets. Wallets will be flushed first.
+void ShutdownWallets();
+
+//! Detach all wallets
```
Same as above.

promag commented on src/wallet/init.h:
```diff
@@ -22,4 +30,15 @@ bool WalletVerify();
 //! Load wallet databases.
 bool InitLoadWallet();
 
+//! Complete startup of wallets
```
Same as above.

promag commented on src/wallet/init.h:
```diff
@@ -6,14 +6,22 @@
 //! Wallets parameter interaction
 bool WalletParameterInteraction();
 
+//! Register wallet RPCs
```
Missing period?

promag replied:
Fix above too.

jnewbery commented:
Re: naming of unload/close/free/detach wallet function. I want the wallet loading/unloading RPC verbs to be antonyms so it's obvious they're carrying out opposite actions. I started with load/unload, but dropped that to avoid any ambiguity with adding keys and addresses to an already open wallet. attach/detach is bad for the reason Russ mentioned. Is everyone happy with open/close?

(I guess the name of the function doesn't need to map exactly to the name of the RPC, but I think it makes sense to do that if we can)

ryanofsky commented:
open/close or load/unload is what I would call it

jnewbery committed:
[wallet] Add StartWallets() function to wallet/init.cpp

jnewbery committed:
[wallet] Rename InitLoadWallet() to OpenWallets()

Rationale:
- this init function can now open multiple wallets (hence
  Wallet->Wallets)
- This is named as the antonym to CloseWallets(), which carries out the
  opposite action.

jnewbery committed:
[wallet] Rename WalletVerify() to VerifyWallets()

This function can now verify multiple wallets.

jnewbery commented:
I've addressed the comments and changed names a little. @ryanofsky do you mind taking a look?

ryanofsky reviewed (commented):
utACK ca94bf2c39b06f1ed643c24dfcd78ef5af00562a. Note that what you're calling flush, I'm calling stop in #10973, and what you're calling stop, I'm calling shutdown. But I'm not too worried about the names.

#10973 links: [declarations](https://github.com/ryanofsky/bitcoin/blob/pr/wipc-sep/src/ipc/interfaces.h), [definitions](https://github.com/ryanofsky/bitcoin/blob/pr/wipc-sep/src/ipc/local/bitcoin-walletd.cpp)

ryanofsky commented:
Also just a suggestion, but this might be easier to review if it were just 2 commits, one commit for RPC stuff and one commit for the start/stop/flush stuff, so you could see the related parts together.

jnewbery commented:
Thanks @ryanofsky .

Other reviewers/maintainers - I'm happy to rebase/squash as Russ suggests if that makes things easier to review. Please let me know if you want that.

MeshCollider commented:
utACK https://github.com/bitcoin/bitcoin/commit/ca94bf2c39b06f1ed643c24dfcd78ef5af00562a

TheBlueMatt commented on src/init.cpp:
```diff
@@ -44,7 +44,6 @@
 #include "validationinterface.h"
 #ifdef ENABLE_WALLET
 #include "wallet/init.h"
-#include "wallet/wallet.h"
```
Holy fuck yes, kill the imports.

TheBlueMatt reviewed (commented):
utACK ca94bf2c39b06f1ed643c24dfcd78ef5af00562a. I'm willing to bet this materially reduces the memory usage/compile time of init.cpp due to the wallet.h import removal.

MarcoFalke commented on src/init.cpp:
```diff
@@ -1567,7 +1559,7 @@ bool AppInitMain(boost::thread_group& threadGroup, CScheduler& scheduler)
     // ********************************************************* Step 8: load wallet
 #ifdef ENABLE_WALLET
-    if (!InitLoadWallet())
+    if (!OpenWallets())
```
Can you add the rationale for rename in the commit message? Why not keep it?

MarcoFalke replied:
@jnewbery Or add it to the OP, so it can be seen in the merge commit.

jnewbery replied:
done!

MarcoFalke commented on src/wallet/init.h:
```diff
@@ -8,18 +8,36 @@
+
 //! Responsible for reading and validating the -wallet arguments and verifying the wallet database.
 //  This function will perform salvage on the wallet if requested, as long as only one wallet is
 //  being loaded (CWallet::ParameterInteraction forbids -salvagewallet, -zapwallettxes or -upgradewallet with multiwallet).
```
Comment should say "WalletParameterInteraction"

jnewbery replied:
added as trivial commit

maflcko reviewed (commented):
utACK. Is mostly move-only

jnewbery committed:
[trivial] fixup comment for VerifyWallets()

jnewbery commented:
@MarcoFalke - nits addressed. Should be ready for merge if Travis agrees.

MarcoFalke commented:
re-utACK commit-by-commit 5d2a3995e7035b3607a11660a2c8330a548f733d
